[analyzer] Fix StackAddrEscapeChecker crash on temporary object fields

steakhal opened the pull request:
Basically, the issue was that we should have unwrap the base region before we special handle temp object regions.

Fixes https://github.com/llvm/llvm-project/issues/66221

I also decided to add some extra range information to the diagnostics to make it consistent with the other reporting path.

DonatNagyE commented on clang/lib/StaticAnalyzer/Checkers/StackAddrEscapeChecker.cpp:
```diff
@@ -398,7 +400,7 @@ void StackAddrEscapeChecker::checkEndFunction(const ReturnStmt *RS,
     }(Referrer->getMemorySpace());
 
     // This cast supposed to succeed.
```
Perhaps explain why will the cast succeed (or, alternatively, just remove this meaningless comment).

steakhal replied:
I was thinking about it once you raised this comment.
To me, to have a proper diagnostic we need something that we can name here; and they are usually variables.

Now, if we check and bail out if it's not a VarRegion, we would not get notified if we suddenly have some other value that we could cover but we don't. It took a long time to even hit this crash, so I don't think they are too frequent.

So the question is which we prefer:
 - Crash (possible in the long future) to get notified or
 - Silently ignore these unthought, interesting cases for which we didn't account for
 
 If this would be a mission-critical software I'd vote for (2), but now I'm a bit hesitant in between the two options.
 
 Maybe, given that this is a core checker, we should lean towards gracefully handling this. WDYT?

DonatNagyE replied:
I think the best would be either a comment that explains _why_ we expect a `VarRegion` here (e.g. it's already checked in function `fooBar()` etc.) or if possible, then a crash-less logic (converting an earlier `isa`-like check + this `cast` into a single `getAs`).

I don't think that we should add defensive/paranoid checks; if this logic is broken by some change in the distant future, then a fixable crash is better than silent degradation.

steakhal replied:
> I think the best would be either a comment that explains _why_ we expect a `VarRegion` here (e.g. it's already checked in function `fooBar()` etc.) or if possible, then a crash-less logic (converting an earlier `isa`-like check + this `cast` into a single `getAs`).

There is nothing that would directly imply this.
It's like, ATM I don't know of any other situations where it could be not a `VarRegion`, and so far it didn't appear in practice that would disprove this.

I'm not exactly sure if a comment like that would suffice. Would you accept it? And do you think it's necessary?

Another alternative could be a combination:

```c++
if (!isa<VarRegion>(Referrer)) {
  assert(false && "We should really only have VarRegions here");
  continue;
}
// ...
cast<VarRegion>(Referrer)...
```

NagyDonat reviewed (approved):
Nice bugfix; I have only one tangential remark.

steakhal commented:
Let me know if you are still okay with the latest change. @DonatNagyE @Xazax-hun

DonatNagyE commented:
LGTM.

steakhal committed:
[analyzer] Fix StackAddrEscapeChecker crash on temporary object fields

Basically, the issue was that we should have unwrap the base region
before we special handle temp object regions.

Fixes https://github.com/llvm/llvm-project/issues/66221

steakhal commented:
Fixed tests for Windows, to use `size_t` for operator new.
I'll land this if premerge-tests pass.
